File: android/src/onramp/java/com/reactnativestripesdk/OnrampErrors.kt
@file:OptIn(ExperimentalCryptoOnramp::class)

package com.reactnativestripesdk

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.reactnativestripesdk.utils.ErrorType
import com.stripe.android.core.exception.StripeException
import com.stripe.android.crypto.onramp.ExperimentalCryptoOnramp
import com.stripe.android.crypto.onramp.exception.CryptoOnrampApiException
import com.stripe.android.crypto.onramp.exception.StripeCryptoOnrampError

internal fun createOnrampFailedError(error: Throwable): WritableMap =
  createOnrampError(ErrorType.Failed.toString(), error)

internal fun createOnrampNotConfiguredError(): WritableMap =
  createOnrampMessageError(
    ErrorType.Failed.toString(),
    "Onramp is not configured.",
  )

private fun createOnrampError(
  code: String,
  error: Throwable,
): WritableMap {
  val apiException =
    error as? CryptoOnrampApiException
      ?: return createNonApiOnrampError(code, error)

  val stripeError = apiException.stripeError
  val onrampErrorType = apiException.toOnrampErrorType()
  val apiErrorContext = apiException.apiErrorContext

  return createOnrampErrorMap(
    code = code,
    message = apiException.userMessage,
    localizedMessage = apiException.localizedMessage,
    declineCode = stripeError?.declineCode,
    type = apiErrorContext.apiErrorType ?: stripeError?.type,
    stripeErrorCode = apiException.code,
  ) {
    putCommonOnrampFields(apiException, onrampErrorType)
    putString("reason", apiErrorContext.reason)
    putString("requestId", apiErrorContext.requestId)
    putString("apiErrorCode", apiErrorContext.apiErrorCode)
    putString("apiErrorType", apiErrorContext.apiErrorType)
    putString("apiErrorMessage", apiErrorContext.apiErrorMessage)
    putString("apiUserMessage", apiErrorContext.apiUserMessage)
  }
}

private fun createNonApiOnrampError(
  code: String,
  error: Throwable,
): WritableMap {
  val onrampError =
    error as? StripeCryptoOnrampError
      ?: return createGenericError(code, error)

  return createOnrampErrorMap(
    code = code,
    message = onrampError.userMessage,
    localizedMessage = error.localizedMessage,
    stripeErrorCode = onrampError.code,
  ) {
    putCommonOnrampFields(onrampError, onrampError.toOnrampErrorType())
  }
}

private fun WritableMap.putCommonOnrampFields(
  error: StripeCryptoOnrampError,
  onrampErrorType: String? = null,
) {
  onrampErrorType?.let {
    putString("onrampErrorType", it)
  }
  putString("developerMessage", error.developerMessage)
  putString("userMessage", error.userMessage)
  putString("docUrl", error.docUrl)
}

private fun createOnrampMessageError(
  code: String,
  message: String?,
): WritableMap =
  createOnrampErrorMap(code = code, message = message)

private fun CryptoOnrampApiException.toOnrampErrorType(): String =
  if (code == "link_failed_to_attest_request") {
    "AppAttestationError"
  } else {
    "UncategorizedApiError"
  }

private fun StripeCryptoOnrampError.toOnrampErrorType(): String? =
  if (code == "app_attestation_unavailable") {
    "AppAttestationUnavailableError"
  } else {
    null
  }

private fun createGenericError(
  code: String,
  error: Throwable,
): WritableMap {
  val stripeError = (error as? StripeException)?.stripeError

  return createOnrampErrorMap(
    code = code,
    message = error.message,
    localizedMessage = error.localizedMessage,
    declineCode = stripeError?.declineCode,
    type = stripeError?.type,
    stripeErrorCode = stripeError?.code,
  )
}

private fun createOnrampErrorMap(
  code: String,
  message: String?,
  localizedMessage: String? = message,
  declineCode: String? = null,
  type: String? = null,
  stripeErrorCode: String? = null,
  configure: (WritableMap.() -> Unit)? = null,
): WritableMap {
  val map: WritableMap = Arguments.createMap()
  val details: WritableMap = Arguments.createMap()

  details.putString("code", code)
  details.putString("message", message)
  details.putString("localizedMessage", localizedMessage)
  details.putString("declineCode", declineCode)
  details.putString("type", type)
  details.putString("stripeErrorCode", stripeErrorCode)
  configure?.invoke(details)

  map.putMap("error", details)
  return map
}

File: ios/OnrampErrors.swift
//
//  OnrampErrors.swift
//  stripe-react-native
//
//  Created by Michael Liberatore on 5/28/26.
//

#if canImport(StripeCryptoOnramp)
import Foundation
@_spi(CryptoOnrampAlpha) import StripeCryptoOnramp

enum OnrampErrors {

    /// Creates a React Native error envelope for a failed Crypto Onramp operation.
    ///
    /// Typed Crypto Onramp errors include onramp-specific diagnostics in addition to the standard error fields.
    static func createFailedError(_ error: Swift.Error) -> NSDictionary {
        return createOnrampError(code: ErrorType.Failed, error: error)
    }

    /// Creates a React Native error envelope for calls made before Crypto Onramp has been configured.
    static func createNotConfiguredError() -> NSDictionary {
        return createOnrampErrorMap(
            code: ErrorType.Failed,
            message: "Onramp is not configured."
        )
    }

    private static func createOnrampError(code: String, error: Swift.Error) -> NSDictionary {
        if let apiError = error as? StripeCryptoOnrampAPIError {
            return createOnrampAPIErrorMap(
                code: code,
                onrampErrorType: onrampErrorType(for: apiError),
                error: apiError
            )
        }

        if let onrampError = error as? StripeCryptoOnrampError {
            return createOnrampErrorMap(
                code: code,
                message: onrampError.userMessage,
                localizedMessage: onrampError.localizedDescription,
                stripeErrorCode: onrampError.code,
                additionalFields: commonOnrampFields(
                    error: onrampError,
                    onrampErrorType: onrampErrorType(for: onrampError)
                )
            )
        }

        return Errors.createError(code, error)
    }

    private static func createOnrampAPIErrorMap(
        code: String,
        onrampErrorType: String,
        error: StripeCryptoOnrampAPIError
    ) -> NSDictionary {
        return createOnrampErrorMap(
            code: code,
            message: error.userMessage,
            localizedMessage: error.localizedDescription,
            type: error.type,
            stripeErrorCode: error.code,
            additionalFields: commonOnrampFields(error: error, onrampErrorType: onrampErrorType).merging([
                "reason": error.reason,
                "requestId": error.requestID,
                "apiErrorCode": error.code,
                "apiErrorType": error.type,
                "apiErrorMessage": error.apiMessage,
                "apiUserMessage": error.apiUserMessage,
            ]) { _, new in new }
        )
    }

    private static func commonOnrampFields(
        error: StripeCryptoOnrampError,
        onrampErrorType: String? = nil
    ) -> [String: Any?] {
        var fields: [String: Any?] = [
            "developerMessage": error.developerMessage,
            "userMessage": error.userMessage,
            "docUrl": error.docURL?.absoluteString,
        ]

        if let onrampErrorType {
            fields["onrampErrorType"] = onrampErrorType
        }

        return fields
    }

    private static func onrampErrorType(for error: StripeCryptoOnrampAPIError) -> String {
        if error.code == "link_failed_to_attest_request" {
            return "AppAttestationError"
        }

        return "UncategorizedApiError"
    }

    private static func onrampErrorType(for error: StripeCryptoOnrampError) -> String? {
        if error.code == "app_attestation_unavailable" {
            return "AppAttestationUnavailableError"
        }

        return nil
    }

    private static func createOnrampErrorMap(
        code: String,
        message: String?,
        localizedMessage: String? = nil,
        declineCode: String? = nil,
        type: String? = nil,
        stripeErrorCode: String? = nil,
        additionalFields: [String: Any?] = [:]
    ) -> NSDictionary {
        let fields: [String: Any?] = [
            "code": code,
            "message": message,
            "localizedMessage": localizedMessage ?? message,
            "declineCode": declineCode,
            "stripeErrorCode": stripeErrorCode,
            "type": type,
        ].merging(additionalFields) { _, new in new }

        return ["error": fields.mapValues(nullIfNil)]
    }

    private static func nullIfNil(_ value: Any?) -> Any {
        return value ?? NSNull()
    }
}
#endif
